##### Austin Hancock

## MSDS 7337 - Section 401
## Homework - #6
[Data Science @ Southern Methodist University](https://datascience.smu.edu/)

### Table of Contents
* [Description](#Description)
* [Tools](#Tools)
* [Question-1](#Question-1)
* [Question-2](#Question-2)

### <a name="Description"></a>Description
For the HW-6 assignment I will be addressing the following:

    - Evaluate the text similarity of book titles from Amazon
    - Evaluate the text similarity of book titles from a major search engine

### <a name="Tools"></a>Tools

In [116]:
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import nltk; print("NLTK", nltk.__version__)
import bs4; print("bs4", bs4.__version__)
from bs4 import BeautifulSoup
import urllib
from urllib import request
import re; print("re", re.__version__)
from nltk.tokenize import word_tokenize
import sklearn; print("sklearn", sklearn.__version__)
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfTransformer
from sklearn.metrics.pairwise import cosine_similarity
from scipy import sparse

Windows-10-10.0.17134-SP0
Python 3.6.3 |Anaconda custom (64-bit)| (default, Oct 15 2017, 03:27:45) [MSC v.1900 64 bit (AMD64)]
NLTK 3.2.4
bs4 4.6.0
re 2.2.1
sklearn 0.19.1


### <a name="Question-1"></a>Question-1

Evaluate the text similarity of Amazon book search results by doing the following: 

    - Do a book search on Amazon and manually copy the top 24 titles
    - Run a text-similarity measure and compare each of the book titles, pairwise, to every other one
    - Find the two titles which are most similar/dissimilar and where they rank amongst the top results

For this project I will be compiling a list of books using the search criteria of 'asteroids'.
The first step is to pass this search criteria into the Amazon search control and specifiy that the search should only return results from 'books'.
Below is the list of the top 24 titles returned from the search.

In [9]:
book_list = [
    "Asteroids: Astronomical and Geological Bodies (Cambridge Planetary Science)",
    "Asteroids IV (Space Science Series)",
    "Asteroids: Relics of Ancient Time",
    "Asteroid Mining 101: Wealth for the New Space Economy",
    "Asteroid Hunters (TED Books)",
    "Impact!: Asteroids and the Science of Saving the World (Scientists in the Field Series)",
    "Asteroid Goddesses: The Mythology, Psychology, and Astrology of the Re-Emerging Feminine",
    "Asteroid Archetypes: A Primer",
    "Planets & Asteroids: Relationships in Conjunction",
    "Kiss My Asteroid: Galaxa Warriors (Paranormal Dating Agency Book 14)",
    "Asteroids & Meteoroids (Our Galaxy)",
    "Asteroid Aftermath (Post Apocalyptic Survival Book 2)",
    "Dawn of Small Worlds: Dwarf Planets, Asteroids, Comets (Astronomers' Universe)",
    "The Silent War: Book III of The Asteroid Wars (The Grand Tour 12)",
    "Asteroids: Prospective Energy and Material Resources",
    "Mechanics of the future: Asteroids",
    "The Asteroid Ephemeris 1900 to 2050",
    "A Is for Asteroids, Z Is for Zombies: A Bedtime Book about the Coming Apocalypse",
    "The Wailing Asteroid",
    "Comets, Meteors, and Asteroids",
    "The Wailing Asteroid: A Science Fiction Story",
    "Catching Stardust: Comets, Asteroids and the Birth of the Solar System (Bloomsbury Sigma)",
    "Asteroids: The Atari 2600 Game Journal",
    "Asteroids III (Space Science Series)"
]

In [22]:
# Tokenize titles

book_list_tokenized = []

for book in book_list:
    text = word_tokenize(book)
    book_list_tokenized.append(text)

In [28]:
# Vectorize the titles using CountVectorizer

count_vectorizer = CountVectorizer()
count_vectorizer.fit_transform(book_list)

<24x105 sparse matrix of type '<class 'numpy.int64'>'
	with 166 stored elements in Compressed Sparse Row format>

In [30]:
# Print a list of word requencies

print(count_vectorizer.vocabulary_)

{'asteroids': 15, 'astronomical': 18, 'and': 10, 'geological': 46, 'bodies': 23, 'cambridge': 26, 'planetary': 68, 'science': 79, 'iv': 54, 'space': 86, 'series': 81, 'relics': 76, 'of': 65, 'ancient': 9, 'time': 93, 'asteroid': 14, 'mining': 61, '101': 0, 'wealth': 101, 'for': 41, 'the': 92, 'new': 64, 'economy': 34, 'hunters': 49, 'ted': 91, 'books': 25, 'impact': 51, 'saving': 78, 'world': 102, 'scientists': 80, 'in': 52, 'field': 40, 'goddesses': 47, 'mythology': 63, 'psychology': 73, 'astrology': 16, 're': 74, 'emerging': 35, 'feminine': 38, 'archetypes': 13, 'primer': 71, 'planets': 69, 'relationships': 75, 'conjunction': 30, 'kiss': 56, 'my': 62, 'galaxa': 43, 'warriors': 99, 'paranormal': 67, 'dating': 31, 'agency': 8, 'book': 24, '14': 2, 'meteoroids': 59, 'our': 66, 'galaxy': 44, 'aftermath': 7, 'post': 70, 'apocalyptic': 12, 'survival': 89, 'dawn': 32, 'small': 84, 'worlds': 103, 'dwarf': 33, 'comets': 28, 'astronomers': 17, 'universe': 96, 'silent': 83, 'war': 98, 'iii': 50

In [37]:
# Create tfidf matrix
tf_matrix = count_vectorizer.transform(book_list).toarray()
tfidfTran = TfidfTransformer(norm="l2")
tfidfTran.fit(tf_matrix)
tfidf_matrix = tfidfTran.transform(tf_matrix)
print(tfidf_matrix.shape)

(24, 105)


This matrix contains the 105 unique words for all of the 24 book titles.

Next, we calulate the pairwise cosine similarity of the titles.

In [38]:
cos_similarity_matrix = (tfidf_matrix * tfidf_matrix.T).toarray()
print(cos_similarity_matrix)

[[1.         0.15538444 0.03953647 0.         0.         0.1373826
  0.05615971 0.         0.03847733 0.         0.0417603  0.
  0.026376   0.         0.11338328 0.04457936 0.         0.02058197
  0.         0.16229809 0.09909496 0.07855977 0.03541959 0.16143223]
 [0.15538444 1.         0.05674252 0.14271707 0.         0.23940094
  0.         0.         0.05522244 0.         0.05993414 0.
  0.03785469 0.         0.0498646  0.06398004 0.         0.02953912
  0.         0.07137674 0.14222051 0.03454964 0.05083399 0.68655767]
 [0.03953647 0.05674252 1.         0.         0.         0.0923783
  0.06506041 0.         0.05031352 0.         0.05460638 0.
  0.10365013 0.06134717 0.04543196 0.17518411 0.         0.02691329
  0.         0.06503181 0.         0.09460056 0.04631518 0.05895102]
 [0.         0.14271707 0.         1.         0.05603497 0.08691514
  0.09488502 0.06726903 0.         0.03411285 0.         0.04615735
  0.         0.11830396 0.         0.0549413  0.0909456  0.1896449
  0.

This matrix is a bit dense to be able to compare visually. Using the below code, we make this dense matrix more readable.

Below, we are converting that very hard to read dense matrix into a sparse one. This helps in two ways. First, it provides a clear representation of what two vectors are being compared. The second way in which it helps is that it removes all 0 values, which are vectors that are not similar at all.

In [115]:
# Convert the dense matrix into a sparse one and sort
sparse_matrix = sparse.csr_matrix(cos_similarity_matrix)
print(sparse_matrix.sorted_indices())

  (0, 0)	0.9999999999999998
  (0, 1)	0.15538444330493628
  (0, 2)	0.03953647239076911
  (0, 5)	0.13738260333189412
  (0, 6)	0.0561597091669565
  (0, 8)	0.038477330487822196
  (0, 10)	0.041760298462173294
  (0, 12)	0.026376004609617705
  (0, 14)	0.11338328012795557
  (0, 15)	0.044579362061319724
  (0, 17)	0.020581969867405563
  (0, 19)	0.162298086636718
  (0, 20)	0.09909496141958725
  (0, 21)	0.07855977176005863
  (0, 22)	0.03541959014363023
  (0, 23)	0.16143223140481372
  (1, 0)	0.15538444330493628
  (1, 1)	0.9999999999999999
  (1, 2)	0.05674251586071726
  (1, 3)	0.14271707072810455
  (1, 5)	0.2394009372568094
  (1, 8)	0.055222441544710586
  (1, 10)	0.059934138140038014
  (1, 12)	0.037854688832912084
  (1, 14)	0.04986459765657751
  :	:
  (22, 16)	0.0541479279141726
  (22, 17)	0.05586980972168168
  (22, 18)	0.10115751881340787
  (22, 19)	0.05826012822868739
  (22, 20)	0.059248659530618523
  (22, 21)	0.1024926374437245
  (22, 22)	0.9999999999999998
  (22, 23)	0.05281252303581939
  (23, 0

In [142]:
# This output shows the cosine similarities between the second book in the list compared to all the books in the list
print(sparse_matrix[1]) 

  (0, 0)	0.15538444330493628
  (0, 1)	0.9999999999999999
  (0, 2)	0.05674251586071726
  (0, 3)	0.14271707072810455
  (0, 5)	0.2394009372568094
  (0, 8)	0.055222441544710586
  (0, 10)	0.059934138140038014
  (0, 12)	0.037854688832912084
  (0, 14)	0.04986459765657751
  (0, 15)	0.06398004186675199
  (0, 17)	0.029539123776726663
  (0, 19)	0.07137673898161397
  (0, 20)	0.14222051387115958
  (0, 21)	0.03454963912128019
  (0, 22)	0.05083399033785032
  (0, 23)	0.6865576669300361


Most similar are titles 2 and 24:
    
    - "Asteroids IV (Space Science Series)"
    - "Asteroids III (Space Science Series)"

There are many titles that have a cosine similarity score of 0, so we pick the two highest ranked below:

    - "Asteroids: Astronomical and Geological Bodies (Cambridge Planetary Science)"
    - "Asteroid Mining 101: Wealth for the New Space Economy"

No surpise with these results. The two most similar titles are books that appear in the same series and only differ in their number within that series. As for the disimilar results, we had many titles that contained no words in common. In terms of rankings within the entire 24 book list that we compiled, the most similar titles were ranked 2nd and 24th, while the two dissimilar titles were ranked 1st and 4th.

### <a name="Question-2"></a>Question-2

Evaluate the results from question 1 compared to a major search engine

    - Compare a book title from the book list with the 1st and 20th results returned from a major search engine.

The book title from the list in question 1 that I will be using is "Asteroid Mining 101: Wealth for the New Space Economy", since it is one of my favorites. Below are the 1st and 20th search results produced from Google for that book title.

In [136]:
book_list_Google = [
    "Asteroid Mining 101: Wealth for the New Space Economy", # Book from question 1
    "Asteroid Mining 101: Wealth for the New Space Economy: John S ...", # Top result from Google
    "The asteroid miner's guide to the galaxy - Daily Herald" #20th result from Google
]

Now we run our same method and output the similarity scores of each.

In [137]:
# Tokenize titles
book_list_Google_tokenized = []

for book in book_list_Google:
    text = word_tokenize(book)
    book_list_Google_tokenized.append(text)
    
# Vectorize the titles using CountVectorizer
count_vectorizer = CountVectorizer()
count_vectorizer.fit_transform(book_list_Google)

# Create tfidf matrix
tf_matrix = count_vectorizer.transform(book_list_Google).toarray()
tfidfTran = TfidfTransformer(norm="l2")
tfidfTran.fit(tf_matrix)
tfidf_matrix = tfidfTran.transform(tf_matrix)

# Pairwise
cos_similarity_matrix = (tfidf_matrix * tfidf_matrix.T).toarray()
print(cos_similarity_matrix)

[[1.         0.90883417 0.17260796]
 [0.90883417 1.         0.15687201]
 [0.17260796 0.15687201 1.        ]]


With only a 3x3 matrix, we do not need to create a sparse matrix in order to compare. Above we can see that the book from question 1 is more similar to the first Google result than to the 20th Google result. Given that the top result was the name of the book along with a partial of the authors name, it is no surpise that the similarity between it and the input value was very high. While both links returned from Google were indeed about the book in question, the first was a link to Amazon for the book while the second was an article written about the book. Clearly, Google is presenting the most pertinent information here as someone who searches for a book by its title probably wants to purchase it and not mearly read an article reviewing it.